File: lib/features/grievance/domain/entities/grievance.dart
class Grievance {
  final String id;
  final String? userId; // null if anonymous
  final String title;
  final String description;
  final String category;
  final String status;
  final String priority;
  final bool isAnonymous;
  final List<String> attachments;
  final String? assignedTo;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? resolvedAt;
  final List<GrievanceUpdate> updates;

  Grievance({
    required this.id,
    this.userId,
    required this.title,
    required this.description,
    required this.category,
    required this.status,
    required this.priority,
    required this.isAnonymous,
    this.attachments = const [],
    this.assignedTo,
    required this.createdAt,
    required this.updatedAt,
    this.resolvedAt,
    this.updates = const [],
  });

  factory Grievance.fromJson(Map<String, dynamic> json) {
    return Grievance(
      id: json['id'] as String,
      userId: json['userId'] as String?,
      title: json['title'] as String,
      description: json['description'] as String,
      category: json['category'] as String,
      status: json['status'] as String,
      priority: json['priority'] as String,
      isAnonymous: json['isAnonymous'] as bool? ?? false,
      attachments: (json['attachments'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      assignedTo: json['assignedTo'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      resolvedAt: json['resolvedAt'] != null
          ? DateTime.parse(json['resolvedAt'] as String)
          : null,
      updates: (json['updates'] as List<dynamic>?)
              ?.map((e) => GrievanceUpdate.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'title': title,
      'description': description,
      'category': category,
      'status': status,
      'priority': priority,
      'isAnonymous': isAnonymous,
      'attachments': attachments,
      'assignedTo': assignedTo,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'resolvedAt': resolvedAt?.toIso8601String(),
      'updates': updates.map((e) => e.toJson()).toList(),
    };
  }

  Grievance copyWith({
    String? id,
    String? userId,
    String? title,
    String? description,
    String? category,
    String? status,
    String? priority,
    bool? isAnonymous,
    List<String>? attachments,
    String? assignedTo,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? resolvedAt,
    List<GrievanceUpdate>? updates,
  }) {
    return Grievance(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      title: title ?? this.title,
      description: description ?? this.description,
      category: category ?? this.category,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      isAnonymous: isAnonymous ?? this.isAnonymous,
      attachments: attachments ?? this.attachments,
      assignedTo: assignedTo ?? this.assignedTo,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      resolvedAt: resolvedAt ?? this.resolvedAt,
      updates: updates ?? this.updates,
    );
  }
}

class GrievanceUpdate {
  final String id;
  final String grievanceId;
  final String message;
  final String updatedBy;
  final String? updatedByRole;
  final DateTime createdAt;

  GrievanceUpdate({
    required this.id,
    required this.grievanceId,
    required this.message,
    required this.updatedBy,
    this.updatedByRole,
    required this.createdAt,
  });

  factory GrievanceUpdate.fromJson(Map<String, dynamic> json) {
    return GrievanceUpdate(
      id: json['id'] as String,
      grievanceId: json['grievanceId'] as String,
      message: json['message'] as String,
      updatedBy: json['updatedBy'] as String,
      updatedByRole: json['updatedByRole'] as String?,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'grievanceId': grievanceId,
      'message': message,
      'updatedBy': updatedBy,
      'updatedByRole': updatedByRole,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}

File: backend/src/models/Grievance.js
const mongoose = require('mongoose');

const grievanceSchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: function () {
            return !this.isAnonymous;
        }
    },
    anonymousId: {
        type: String,
        required: function () {
            return this.isAnonymous;
        },
        index: true
    },
    title: {
        type: String,
        required: [true, 'Title is required'],
        trim: true,
        minlength: [5, 'Title must be at least 5 characters'],
        maxlength: [200, 'Title cannot exceed 200 characters']
    },
    description: {
        type: String,
        required: [true, 'Description is required'],
        trim: true,
        minlength: [20, 'Description must be at least 20 characters'],
        maxlength: [2000, 'Description cannot exceed 2000 characters']
    },
    category: {
        type: String,
        required: [true, 'Category is required'],
        enum: [
            'Academic Issues',
            'Infrastructure',
            'Faculty Behavior',
            'Examination',
            'Library',
            'Hostel',
            'Canteen',
            'Transportation',
            'Ragging/Harassment',
            'Discrimination',
            'Safety & Security',
            'Administrative',
            'Other'
        ]
    },
    status: {
        type: String,
        enum: ['pending', 'under_review', 'in_progress', 'resolved', 'rejected'],
        default: 'pending',
        index: true
    },
    priority: {
        type: String,
        enum: ['low', 'medium', 'high', 'urgent'],
        default: 'medium',
        index: true
    },
    isAnonymous: {
        type: Boolean,
        default: false
    },
    isEmergency: {
        type: Boolean,
        default: false
    },
    attachments: [{
        filename: String,
        originalName: String,
        mimetype: String,
        size: Number,
        path: String,
        uploadedAt: {
            type: Date,
            default: Date.now
        }
    }],
    assignedTo: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User'
    },
    updates: [{
        message: {
            type: String,
            required: true
        },
        updatedBy: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User',
            required: true
        },
        updatedByRole: {
            type: String,
            enum: ['student', 'committee', 'admin']
        },
        createdAt: {
            type: Date,
            default: Date.now
        }
    }],
    resolvedAt: {
        type: Date
    },
    createdAt: {
        type: Date,
        default: Date.now,
        index: true
    },
    updatedAt: {
        type: Date,
        default: Date.now
    }
});

// Update timestamp on save
grievanceSchema.pre('save', function (next) {
    this.updatedAt = Date.now();

    // Set resolvedAt when status changes to resolved
    if (this.isModified('status') && this.status === 'resolved' && !this.resolvedAt) {
        this.resolvedAt = Date.now();
    }

    next();
});

// Compound indexes for efficient queries
grievanceSchema.index({ userId: 1, status: 1, createdAt: -1 });
grievanceSchema.index({ anonymousId: 1, createdAt: -1 });
grievanceSchema.index({ category: 1, status: 1 });
grievanceSchema.index({ priority: 1, status: 1 });

module.exports = mongoose.model('Grievance', grievanceSchema);
